File: Supervised Learning/Decision Tree/coding.R
library(rpart)
library(caret)

data(mtcars)


# Regression

set.seed(42)

sample <- sample(1:nrow(mtcars), 0.8 * nrow(mtcars))
train <- mtcars[sample, ]
test <- mtcars[-sample, ]


reg_model <- rpart(mpg ~ ., data = train)
y_pred <- predict(reg_model, newdata = test)


mse <- mean((test$mpg - y_pred) ^ 2)
r2 <- 1 - sum((test$mpg - y_pred)^2) / sum((test$mpg - mean(test$mpg))^2)

print(paste("MSE:", mse))
print(paste("R-squared:", r2))


# Or using train method

reg_dt_model <- train(mpg ~ ., 
                   data = train,
                   method = "rpart",
                   trControl = trainControl(method = "cv", number = 5),
                   tuneLength = 10)

print(reg_dt_model)

print(reg_dt_model$bestTune)



# Classification

data("iris")

set.seed(42)
sample <- sample(1:nrow(iris), 0.8 * nrow(iris))
train <- iris[sample, ]
test <- iris[-sample, ]

class_model <- rpart(Species ~ ., data = train)
y_pred <- predict(class_model, test, type = "class")

conf_matrix <- confusionMatrix(test$Species, y_pred)

print(conf_matrix)

File: Supervised Learning/Decision Tree/coding.py
from sklearn.datasets import load_iris, load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor
from sklearn.metrics import accuracy_score, mean_squared_error, r2_score

# regression
df = load_breast_cancer()

X, y = df.data, df.target

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

reg_model = DecisionTreeRegressor()
reg_model.fit(X_train, y_train)
y_pred = reg_model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse}")
print(f"R-squared: {r2}")


# Classification

df = load_iris()

X, y = df.data, df.target

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

class_model = DecisionTreeClassifier()
class_model.fit(X_train, y_train)
y_pred = class_model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)

print(f"Accuracy: {accuracy}")
